### PR TITLE
Add card scanning support

### PR DESCRIPTION
**Pull request by davidme-stripe**
## Summary
* Adds card scanning support using iOS 13's Vision framework.
* Removes support for Card.io, which is unmaintained and can't read newer cards.

We're fairly confident in the coverage of this scanner, but we're releasing it as a beta for now. Please let us know if you encounter any bugs or have trouble with a specific type of card.

## Testing
* Tested on iPhone 7, iPhone 11 Plus, and iPad Pro with iOS 13 and iOS 14.

### CODE REVIEW

File: Stripe/STPAddCardViewController.m [hunk 1]
```diff
@@ -202,8 +223,29 @@ - (void)setAlwaysEnableDoneButton:(BOOL)alwaysEnableDoneButton {
...
+- (void)setIsScanning:(BOOL)isScanning {
+    if (_isScanning == isScanning) {
+        return;
+    }
+    _isScanning = isScanning;
+    
+    self.cardHeaderView.button.enabled = !isScanning;
+    NSIndexPath *indexPath = [NSIndexPath indexPathForRow:0 inSection:STPPaymentCardScannerSection];
+    if (isScanning) {
+        [self.tableView insertRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
```
Comment (csabol-stripe) [inline Stripe/STPAddCardViewController.m:235]: this should be wrapped in `[self.tableView beginUpdates/endUpdates]`

File: Stripe/STPAddCardViewController.m [hunk 2]
```diff
@@ -526,11 +581,51 @@ - (void)useShippingAddress:(__unused UIButton *)sender {
...
+        __block NSUInteger i = 0;
+        self.scannerCompleteAnimationTimer = [NSTimer scheduledTimerWithTimeInterval:kSTPCardScanAnimationTime repeats:YES block:^(NSTimer * _Nonnull timer) {
+            i++;
+            STPPaymentMethodCardParams *newParams = [[STPPaymentMethodCardParams alloc] init];
+            if (i < [cardParams.number length]) {
+                newParams.number = [cardParams.number substringToIndex:i];
+            } else {
+                newParams.number = cardParams.number;
+            }
+            self.paymentCell.paymentField.cardParams = newParams;
```
Comment (csabol-stripe) [inline Stripe/STPAddCardViewController.m:616]: maybe disable user interaction during this?

File: Stripe/STPCardScanner.m [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+
+- (void)setupCamera {
+    __weak typeof(self) weakSelf = self;
+    self.textRequest = [[VNRecognizeTextRequest alloc] initWithCompletionHandler:^(VNRequest * _Nonnull request, NSError * _Nullable error) {
+        __strong typeof(self) strongSelf = weakSelf;
+        if (!strongSelf.isScanning) {
+            return;
+        }
+        if (error) {
+            NSLog(@"Text recognition error.");
```
Comment (csabol-stripe) [inline Stripe/STPCardScanner.m:141]:
```suggestion
```

File: Stripe/STPCardScanner.m [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+    AVCaptureDevice *captureDevice = [AVCaptureDevice defaultDeviceWithDeviceType:AVCaptureDeviceTypeBuiltInWideAngleCamera mediaType:AVMediaTypeVideo position:AVCaptureDevicePositionBack];
+    self.captureDevice = captureDevice;
+    
+    self.captureSession = [[AVCaptureSession alloc] init];
+    self.captureSession.sessionPreset = AVCaptureSessionPreset1920x1080;
+    
+    NSError *deviceInputError;
+    AVCaptureDeviceInput *deviceInput = [[AVCaptureDeviceInput alloc] initWithDevice:captureDevice error:&deviceInputError];
+    if (deviceInputError) {
+        NSLog(@"Failed to create camera device input.");
```
Comment (csabol-stripe) [inline Stripe/STPCardScanner.m:157]:
```suggestion
```

File: Stripe/STPCardScanner.m [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+    } else {
+        [self stop];
+        return;
+    }
+    
+    self.videoDataOutputQueue = dispatch_queue_create("com.stripe.CardScanning.VideoDataOutputQueue", nil);
+    self.videoDataOutput = [[AVCaptureVideoDataOutput alloc] init];
+    self.videoDataOutput.alwaysDiscardsLateVideoFrames = YES;
+    [self.videoDataOutput setSampleBufferDelegate:self queue:self.videoDataOutputQueue];
+    [self.videoDataOutput setVideoSettings:@{(id)kCVPixelBufferPixelFormatTypeKey : @(kCVPixelFormatType_420YpCbCr8BiPlanarFullRange)}];
```
Comment (csabol-stripe) [inline Stripe/STPCardScanner.m:173]: maybe a comment on what this means haha?

File: Stripe/STPCardScanner.m [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+    
+    // This improves recognition quality, but means the VideoDataOutput buffers won't match what we're seeing on screen.
+    [[self.videoDataOutput connectionWithMediaType:AVMediaTypeVideo] setPreferredVideoStabilizationMode:AVCaptureVideoStabilizationModeAuto];
+    
+    [self.captureSession startRunning];
+    
+    NSError *lockError;
+    [self.captureDevice lockForConfiguration:&lockError];
+    if (lockError) {
+        NSLog(@"Failed to lock the camera: Our card scanning session won't be zoomed in.");
```
Comment (csabol-stripe) [inline Stripe/STPCardScanner.m:190]:
```suggestion
```

File: Stripe/STPCardScanner.m [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+        return;
+    }
+    self.textRequest.recognitionLevel = VNRequestTextRecognitionLevelAccurate;
+    self.textRequest.usesLanguageCorrection = NO;
+    self.textRequest.regionOfInterest = self.regionOfInterest;
+    VNImageRequestHandler *handler = [[VNImageRequestHandler alloc] initWithCVPixelBuffer:pixelBuffer orientation:self.textOrientation options:@{}];
+    NSError *requestError;
+    [handler performRequests:@[self.textRequest] error:&requestError];
+    if (requestError) {
+        NSLog(@"OCR failed.");
```
Comment (csabol-stripe) [inline Stripe/STPCardScanner.m:214]:
```suggestion
```

File: Stripe/STPCardScanner.m [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+@property (nonatomic, strong) AVCaptureVideoDataOutput *videoDataOutput;
+@property (nonatomic, strong) dispatch_queue_t videoDataOutputQueue;
+
+@property (nonatomic, strong) VNRecognizeTextRequest *textRequest;
+
+@property (atomic) BOOL isScanning;
+@property (atomic) BOOL didTimeout;
+@property (atomic) BOOL timeoutStarted;
+
+@property (atomic) UIDeviceOrientation _deviceOrientation;
```
Comment (csabol-stripe) [inline Stripe/STPCardScanner.m:42]: Is this a namespacing this? If so probably _stp_deviceOrientation is safer

File: Stripe/STPCardScannerTableViewCell.m [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+@interface STPCardScannerTableViewCell()
+
+@property (nonatomic, weak) STPCameraView *cameraView;
+
+@end
+
+@implementation STPCardScannerTableViewCell
+
+static const CGFloat cardSizeRatio = 2.125f/3.370f;
```
Comment (csabol-stripe) [inline Stripe/STPCardScannerTableViewCell.m:23]: comment for this number?

**Comment by aoropeza** [comment]
Amazing, I'm exiting to introduce this new feature to my app.

**Pull request merged**